noir-projects/noir-contracts/contracts/app_subscription_contract/src/main.nr

```diff
@@ -111,16 +111,16 @@ contract AppSubscription {
         AppSubscription::at(context.this_address()).assert_block_number(expiry_block_number).static_enqueue(&mut context);
 
         let mut subscription_note = SubscriptionNote::new(subscriber_address, expiry_block_number, tx_count);
-        if (!is_initialized(subscriber_address)) {
+
+        // is_initialized is an unconstrained function, so we must first convert the state variable to an unconstrained
+        // one to avoid having a mutable reference (to the private context) cross the constrained <> unconstrained
+        // boundary, which is not allowed.
+        let is_initialized = storage.subscriptions.at(subscriber_address).to_unconstrained().is_initialized();
+
+        if (!is_initialized) {
             storage.subscriptions.at(subscriber_address).initialize(&mut subscription_note, true);
         } else {
             storage.subscriptions.at(subscriber_address).replace(&mut subscription_note, true)
         }
     }
-
-    // Compiler bug workaround. You can't call an unconstrained function in another module, unless its from an
-    // unconstained function in your module.
-    unconstrained fn is_initialized(subscriber_address: AztecAddress) -> pub bool {
-        storage.subscriptions.at(subscriber_address).is_initialized()
-    }
 }
```

Comment:
This comment refers to a compiler bug workaround, but the 'issue' was fixed incorrectly. The referenced bug was most likely https://github.com/noir-lang/noir/issues/5008, though the only bug is in diagnosis as the code was still wrong: we cannot call `unconstrained` state variable methods from constrained functions since the object carries a mutable reference (to the context) which cannot cross the const<>unconstr boundary.

The only options are:
 - to create a constrained version of `is_initialized` (e.g. with a historical proof). I understand we don't need to since the following `if` statement handles both scenarios correctly.
 - to make the current `is_initialized()` constrained, which would be semantically wrong since it is not
 - to explicitly opt-in into unconstrained execution, which is odd

Likely a better fourth choice would be to review `PrivateMutable` and its initialization in general, and possibly provide a primitive that performs the above instead of leaving it up to user code. I left things as-is to avoid introducing further changes here.

Reply:
Edit after writing: my brain had to go through this thought process to be sure this was ok. It's maybe worrying the level of thinking needed to validate this snippet.

Whenever I see values being initialised from an unconstrained function, it sets off alarm bells (just because that's how so many under-constrained bugs happen in circuits).
Why isn't `is_initialized` a constrained function? Shouldn't there be constraints to validate that the initialisation nullifier has already been emitted (by checking the nullifier tree via the archive tree)? [Edit after writing: the subsequent lines do seem to be constraining this, but it took me a bit to validate].

If it's unconstrained (as it is here), then a malicious user can set `is_initialized` to `true` or `false` at will.

If the user chooses `true`:
Then in the subsequent `if` statement (`if (!is_initialized)`), the `else` clause will trigger and the user can perform `replace`. Does `replace` perform a `read_request` on the storage slot to ensure a note already exists?* Otherwise I could initialise the state this way, and then initialise multiple versions of the same state through repeated calls... Ok I've checked and it does do a read_request. But I had to go looking in the innards of aztec.nr to be sure. So that means it's ok. A previous note must already exist, which means it must have been initialised, which is what we want.

If the user chooses `false`:
Then in the subsequent `if` statement (`if (!is_initialized)`), the `if` clause will trigger. Ok, then it'll initialize and emit the nullifier, so this branch can't be called more than once.



*Remember that conversation we had in that meeting room in London where we talked about naming these methods, and I wanted long, ugly names which exactly described what the method would do? It's because I knew if we didn't do that I'd need to go hunting in aztec.nr to remember what the method actually does under-the-hood every time I read aztec.nr. Lo and behold, I had to go hunting!  :P

Reply:
Thanks for going over this! I fully agree - `is_initialized()` as is is a bad function which should not be part of our API, because it is very hard to use safely. I'd suggest we have:
 - a constrained `is_initialized()` that performs the inclusion check
 - a constrained `initialize_or_replace()` that performs the full snippet from above (unconstrained check plus if-else chain that correctly handles both cases, where the unconstrained part is simply a hint)

I think the fact that this forced us to create the `to_unconstrained` method hints at an underlying issue, which seems to be another positive of the approach followed in this PR. We shouldn't need to do such things.

Reply:
Created https://github.com/AztecProtocol/aztec-packages/pull/6519 to address this